README.md
# syscontainer-tools

## Introduction

**syscontainer-tools** is a fully customized tool,
it is a small auxiliary tool which is expected to work with isulad with hook support,
and provides enhanced functions which is inappropriate to be included in isulad itself.

The project includes two main parts: `syscontainer-tools` and `hooks`.
`syscontainer-tools` is used for dynamically operating on running containers,
and `hooks` is used for executing user defined program at some special timepoint of container's lifecycle.

## Hooks

We provide syscontainer hooks function.

* syscontainer-hooks:
 1. insert block devices added by syscontainer-tools into container when container restarts(prestart state).
 2. insert network interfaces and route rules added by syscontainer-tools into container when container restarts(prestart state).
 3. remove udev rules and leaking network interfaces when container stops(post-stop state).
 4. handling oci relabel for container in prestart and post stop state.

You could use hook spec to customise your hooks.
For detailed information, See [introduction of syscontainer-hooks](hooks/syscontainer-hooks/README.md)

## syscontainer-tools

Basic usage of `syscontainer-tools`:

```
NAME:
   syscontainer-tools - Enhanced tools for IT isulad

USAGE:
   syscontainer-tools [global options] command [command options] [arguments...]

VERSION:
   v0.9
commit: e39c47b1d0403fd133c49db13ab6df7e5d53a21b

COMMANDS:
    add-device          add one or more host devices to container
    add-nic             create a new network interfaces for container
    add-path            add one or more host paths to container
    add-route           add a new network route rule into container
    relabel             relabel rootfs for running SELinux in system container
    remove-device       remove one or more devices from container
    remove-nic          remove a network interface from container
    remove-path         remove one or more paths from container
    remove-route        remove a network route rule from container

GLOBAL OPTIONS:
   --debug                              enable debug output for logging
   --log "/dev/null"                    set the log file path where internal debug information is written
   --log-format "text"                  set the format used by logs ('text' (default), or 'json')
   --syslog-service "unix:///dev/log"   set syslog service
   --help, -h                           show help
   --version, -v                        print the version
```

For usage of each command, you can check with `--help`, e.g. `syscontainer-tools add-device --help`

## Contributions

As this is a fully customized tool, I don't think anyone will be interested in contributing to this project,
but we welcome your contributions. Before contributing, please make sure you understand our needs and
make a communication with us: isulad@openeuler.org.


CHANGELOG
# CHANGELOG

## v0.9 (2017-9-1)

* Add list-nic command and change remove-nic --name paramer form (MR: #93)
* Update config only when container isn't running (MR: #94)
* Add new parameters (--pretty/--filter) about nic/route commands (MR: #95)
* Add list route command (MR: #95)

## v0.8 (2017-8-16)

### general

* change Mtu range (MR: #80)
* remove 'add-path/remove-path' commands from syscontainer-tools (MR: #82)
* network: config net device 'tx/sg/tso' on (MR: #81)
* fix some minor issue(security check, file mode...) (MR: #84)
* fix when supply empty mac address, `add-nic` will panic (MR: #85)
* add '--update-config-only' option for add resources (MR: #89)
* add 'follow-partition' for add/remove device (MR: #86)

## v0.7 (2017-6-29)

### general

* reorginzed the syscontainer-tools code  (MR: #75)
* use`syscontainer-tools` in `syscontainer-tools_wrapper` instead of `isula exec` (MR: #76)
* fix filelock can't lock accross the process issue (MR: #78)

### hooks

* combine 3 hooks binary to `syscontainer-hooks` (MR: #75)

## v0.6 (2017-6-23)

### general

* syscontainer-tools add `add-route & del-nic & del-route` command line. (MR: #71, #68)
* syscontainer-tools add-nic supports network interface config. (MR: #68)
* syscontainer-tools supports fidsk in container. (MR: #70)
* syscontainer-tools could only remove the devices added by syscontainer-tools (MR: #66)

### network-hook

* supports persistent network config for both interfaces and route rules. (MR: #69)

## v0.5 (2017-2-15)

### general

* fix some dir permissions for potential security risk. (MR: #56, #57, #62)
* move empty_systemd.sh to iSulad project. (MR: #60)

### syscontainer-tools

* fix syscontainer-tools subcommand display orders. (MR: #55)
* when iSulad daemon and syscontainer-tools are not in the same cgroup path, device-hook will fail to update cgroup permissions. (MR: #58)
* when containers specify cgroup folder (with `--cgroup-parent` flag), syscontainer-tools will fail to update device cgroup permissions. (MR: #64)

### oci-relabel-hook

* Fix systemd's context type if host use iSulad image's SELinux policy. (MR: #61)

## v0.4 (2017-1-24)

### general

* fix some golang security problems based on review comments

### syscontainer-tools

* Fix find cgroup path issue when syscontainer-tools run in container
* Add upstart support for system container

## v0.3 (2017-1-9)

### general

* use "go+c" function call to take over "exec nsenter"
* add golang 1.5.4 support

### hooks

* device-hook: Update cgroup permission in prestart hook not poststart
* oci-relabel-hook: Use golang to refact oci-relabel-hook

### syscontainer-tools

* Add "--force" flag for `syscontainer-tools add-device`

* can not use fdisk to partition when add a device to a container

## v1.11.2.8.it

### hooks

* device-hook: newly introduced device hook to save config for new added devices.
* oci-relabel-hook: use container's SELinux policy to adapt various images.

### syscontainer-tools

* use device-hook instead of "isula update" to keep resource persistent in container.
* add syslog support, all log will be written to syslog.
* add "make rpm" support in Makefile.

### bug fix

* can not add read-only mount path to container.
* add resource to system container, then use reboot to restart container, then resource can not be found after next reboot.
* validate network interface name in container, error out if it's empty.
* error out when insert network interface to non-existing container.

## v1.11.2.7.it

This is first initial release with lots of new features:

### hooks

* network-hook: move network interface from host to container when container starts.
* oci-relabel-hook: relabel container's rootfs when container starts.

### syscontainer-tools

New Commands:

* add-device: dynamically add new device into running container.
* remove-device: remove device from running container.
* add-path: dynamically bind mount host path to container path.
* remove-path: remove bind mount path from container.
* add-nic: dynamically move physical/virtual network interfaces into container.


network.go
// Copyright (c) Huawei Technologies Co., Ltd. 2018-2019. All rights reserved.
// syscontainer-tools is licensed under the Mulan PSL v2.
// You can use this software according to the terms and conditions of the Mulan PSL v2.
// You may obtain a copy of Mulan PSL v2 at:
//    http://license.coscl.org.cn/MulanPSL2
// THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR
// PURPOSE.
// See the Mulan PSL v2 for more details.
// Description: network interface commands
// Author: zhangwei
// Create: 2018-01-18

// go base main package
package main

import (
	"bytes"
	"encoding/json"
	"os"
	"strings"

	"github.com/urfave/cli"

	"github.com/docker/libnetwork/netutils"
	"github.com/sirupsen/logrus"
	"isula.org/syscontainer-tools/container"
	"isula.org/syscontainer-tools/libnetwork"
	"isula.org/syscontainer-tools/types"
)

var addNicCommand = cli.Command{
	Name:      "add-nic",
	Usage:     "create a new network interface for container",
	ArgsUsage: `<container_id>`,
	Description: `This command is used to create a new network interface in an existing container,
and configure it as you wanted, then attach to specified bridge.
	`,
	Flags: []cli.Flag{
		cli.StringFlag{
			Name:  "type",
			Usage: "set network interface type (veth/eth)",
		},
		cli.StringFlag{
			Name:  "name",
			Usage: "set network interface name: [host:]<container>. for veth type, host could be unset and is random by default. for eth type, host is required",
		},
		cli.StringFlag{
			Name:  "ip",
			Usage: "set ip address. E.g. 172.17.28.2/24",
		},
		cli.StringFlag{
			Name:  "mac",
			Usage: "set mac address. E.g. 00:ff:48:23:e2:bb",
		},
		cli.StringFlag{
			Name:  "bridge",
			Usage: "set bridge name the network interface will attach to, for eth type, bridge cannot be set",
		},
		cli.IntFlag{
			Name:  "mtu",
			Value: 1500,
			Usage: "set mtu",
		},
		cli.IntFlag{
			Name:  "qlen",
			Value: 1000,
			Usage: "set qlen, 1000 by default",
		},
		cli.BoolFlag{
			Name:  "update-config-only",
			Usage: "If this flag is set, will not add network interface to container but update config only",
		},
	},
	Action: func(context *cli.Context) {
		if context.NArg() < 1 {
			fatalf("%s: %q must accept a container-id", os.Args[0], context.Command.Name)
		}

		ctrName := strings.TrimSpace(context.Args()[0])
		if len(ctrName) == 0 {
			fatalf("container-id can't be empty")
		}

		if context.NArg() > 1 {
			fatalf("Don't put container-id in the middle of options")
		}

		hostNicName, ctrNicName, err := parseNicName(context.String("name"))
		if err != nil {
			fatalf("failed to parse network name: %s", context.String("name"))
		}

		ctr, err := container.New(ctrName)
		if err != nil {
			fatalf("failed to get container info: %v", err)
		}
		if ctrNicName == "" {
			fatalf("failed to get container nic name")
		}

		nicConf := &types.InterfaceConf{
			IP:          context.String("ip"),
			Mac:         context.String("mac"),
			Mtu:         context.Int("mtu"),
			Type:        context.String("type"),
			Bridge:      context.String("bridge"),
			Qlen:        context.Int("qlen"),
			CtrNicName:  ctrNicName,
			HostNicName: hostNicName,
		}

		if err := types.ValidNetworkConfig(nicConf); err != nil {
			fatalf("invalid network option: %v", err)
		}

		if nicConf.HostNicName == "" {
			nicConf.HostNicName = nicConf.CtrNicName
			// interface name length must be less than 15
			if len(nicConf.CtrNicName) > 9 {
				nicConf.HostNicName = nicConf.CtrNicName[:9]
			}
			nicConf.HostNicName, err = netutils.GenerateRandomName(nicConf.HostNicName+"_", 5)
			if err != nil {
				fatalf("failed set host nic name %v", err)
			}
		}
		if err := libnetwork.AddNic(ctr, nicConf, context.Bool("update-config-only")); err != nil {
			fatalf("failed to add nic into container: %v", err)
		}
		logrus.Infof("add network interface to container %s successfully", ctrName)
	},
}

var rmNicCommand = cli.Command{
	Name:      "remove-nic",
	Usage:     "remove a network interface from container",
	ArgsUsage: `<container_id>`,
	Description: `This command is used to remove a network interface from an existing container.
	`,
	Flags: []cli.Flag{
		cli.StringFlag{
			Name:  "type",
			Usage: "set network interface type (veth/eth)",
		},
		cli.StringFlag{
			Name:  "name",
			Usage: "set network interface name: [host:]<container>",
		},
	},
	Action: func(context *cli.Context) {
		if context.NArg() < 1 {
			fatalf("%s: %q must accept a container-id", os.Args[0], context.Command.Name)
		}

		ctrName := strings.TrimSpace(context.Args()[0])
		if len(ctrName) == 0 {
			fatalf("container-id can't be empty")
		}

		if context.NArg() > 1 {
			fatalf("Don't put container-id in the middle of options")
		}

		hostNicName, ctrNicName, err := parseNicName(context.String("name"))
		if err != nil {
			fatalf("failed to parse network name: %s", context.String("name"))
		}

		ctr, err := container.New(ctrName)
		if err != nil {
			fatalf("failed to get container info: %v", err)
		}

		nicConf := &types.InterfaceConf{
			CtrNicName:  ctrNicName,
			HostNicName: hostNicName,
			Type:        context.String("type"),
		}

		if err := libnetwork.DelNic(ctr, nicConf); err != nil {
			fatalf("failed to remove nic from container: %v", err)
		}
		logrus.Infof("remove network interface from container %v successfully", ctrName)
	},
}

func parseNicName(name string) (string, string, error) {
	names := strings.SplitN(name, ":", 2)
	if len(names) == 1 {
		return "", name, nil
	}

	return names[0], names[1], nil
}

var updateNicCommand = cli.Command{
	Name:      "update-nic",
	Usage:     "update network interfaces in a container",
	ArgsUsage: `<container_id>`,
	Description: `This command is used to update network interfaces in an existing container.
	`,
	Flags: []cli.Flag{
		cli.StringFlag{
			Name:  "name",
			Usage: "network interface name in container will be updated, must be required",
		},
		cli.StringFlag{
			Name:  "ip",
			Usage: "set ip address. E.g. 172.17.28.2/24",
		},
		cli.StringFlag{
			Name:  "mac",
			Usage: "set mac address. E.g. 00:ff:48:23:e2:bb",
		},
		cli.StringFlag{
			Name:  "bridge",
			Usage: "set bridge name the network interface will attach to",
		},
		cli.IntFlag{
			Name:  "mtu",
			Usage: "set mtu. 0 means keeping old value",
		},
		cli.IntFlag{
			Name:  "qlen",
			Usage: "set qlen, works only on veth type. 0 means keeping old value",
		},
		cli.BoolFlag{
			Name:  "update-config-only",
			Usage: "If this flag is set, will not add network interface to container but update config only",
		},
	},
	Action: func(context *cli.Context) {
		if context.NArg() < 1 {
			fatalf("%s: %q must accept a container-id", os.Args[0], context.Command.Name)
		}

		ctrName := strings.TrimSpace(context.Args()[0])
		if len(ctrName) == 0 {
			fatalf("container-id can't be empty")
		}

		if context.NArg() > 1 {
			fatalf("Don't put container-id in the middle of options")
		}

		ctr, err := container.New(ctrName)
		if err != nil {
			fatalf("failed to get container info: %v", err)
		}

		ctrNicName := context.String("name")
		if ctrNicName == "" {
			fatalf("Network interface name in container must be provided")
		}

		// we use qlen < 0 to tell qlen is not set by user, so set it to -1 here
		var qlen int
		if !context.IsSet("qlen") {
			qlen = -1
		} else {
			qlen = context.Int("qlen")
		}

		nicConf := &types.InterfaceConf{
			IP:         context.String("ip"),
			Mac:        context.String("mac"),
			Mtu:        context.Int("mtu"),
			Bridge:     context.String("bridge"),
			Qlen:       qlen,
			CtrNicName: ctrNicName,
		}

		if err := libnetwork.UpdateNic(ctr, nicConf, context.Bool("update-config-only")); err != nil {
			fatalf("failed to upadte nic in container: %v", err)
		}

		logrus.Infof("update network interface in container %v successfully", ctrName)
	},
}

var listNicCommand = cli.Command{
	Name:      "list-nic",
	Usage:     "list all network interfaces in a container",
	ArgsUsage: `<container_id>`,
	Description: `This command is used to list all network interfaces in an existing container.
	`,
	Flags: []cli.Flag{
		cli.BoolFlag{
			Name:  "pretty, p",
			Usage: "If this flag is set, list nics in pretty json form",
		},
		cli.StringFlag{
			Name:  "filter, f",
			Usage: "Filter output based on conditions provided. E.g. '{\"ip\":\"1.2.3.4/24\", \"Mtu\":1500}'",
		},
	},
	Action: func(context *cli.Context) {
		if context.NArg() < 1 {
			fatalf("%s: %q must accept a container-id", os.Args[0], context.Command.Name)
		}

		ctrName := strings.TrimSpace(context.Args()[0])
		if len(ctrName) == 0 {
			fatalf("container-id can't be empty")
		}

		if context.NArg() > 1 {
			fatalf("Don't put container-id in the middle of options")
		}

		ctr, err := container.New(ctrName)
		if err != nil {
			fatalf("failed to get container info: %v", err)
		}

		filterString := context.String("filter")
		if filterString == "" {
			filterString = "{}"
		}
		var filter = new(types.InterfaceConf)
		if err = json.Unmarshal([]byte(filterString), filter); err != nil {
			fatalf("failed to parse filter: %v", err)
		}

		nics, err := libnetwork.ListNic(ctr, filter)
		if err != nil {
			fatalf("failed to list nic in container: %v", err)
		}
		if nics == nil || len(nics) == 0 {
			logrus.Infof("list network interface in container %q successfully", ctrName)
			return
		}

		nicData, err := json.Marshal(nics)
		if err != nil {
			fatalf("failed to Marshal nic config: %v", err)
		}
		nicBuffer := new(bytes.Buffer)
		if _, err = nicBuffer.Write(nicData); err != nil {
			fatalf("Buffer Write error %v", err)
		}

		if context.Bool("pretty") {
			nicBuffer.Truncate(0)
			if json.Indent(nicBuffer, nicData, "", "\t") != nil {
				fatalf("failed to Indent nic data: %v", err)
			}
		}
		if _, err = nicBuffer.WriteString("\n"); err != nil {
			fatalf("Buffer WriteString error %v", err)
		}
		if _, err = os.Stdout.Write(nicBuffer.Bytes()); err != nil {
			logrus.Errorf("Write nicBuffer.Bytes error: %v", err)
		}
		logrus.Infof("list network interface in container %v successfully", ctrName)
	},
}
